### PR TITLE
Adds Bit-Bots Quintic Walk 

### PR DESCRIPTION
**Pull request by ysims**
Adds Bit-Bots' Quintic Walk to the codebase. Uses splines. Is based off a walk by Team Rhoban.

- [x] Clean code (comments, logs)
- [ ] Documentation in NUbook (https://github.com/NUbots/NUbook/pull/44)
- [x] Works with Keyboard Walk
- [x] Works with RoboCup
- [ ] ~~Works with PS3 Walk~~
- [x] Memory issues fixed

### CODE REVIEW

File: log.txt [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+Installing support::configuration::NetworkConfiguration
+Installing support::NUsight
+Installing platform::darwin::HardwareSimulator
+Installing platform::darwin::SensorFilter
+Installing motion::QuinticWalk
+Installing motion::HeadController
+Installing behaviour::Controller
+Installing behaviour::skills::DirectWalkController
+Installing behaviour::strategy::KeyboardWalk
+[?1049h[22;0;0t[1;24r(B[m[4l[?7h[?1h=
```
Comment (TrentHouliston) [inline log.txt]: What is this doing here :P

**Review by TrentHouliston — DISMISSED** [review]

File: log.txt (hunk 1 above)
Comment (ysims) [reply log.txt]: I thought I deleted that lol
and this is why I dont have a PR for the static walk bc I want to work on multiple computers and I also want to put test stuff in

File: module/motion/QuinticWalk/src/engine/TrajectoryUtils.hpp [hunk 2]
```diff
@@ -0,0 +1,307 @@
...
+
+        /**
+         * Simple typedef for trajectories container
+         */
+        using Trajectories = SplineContainer<SmoothSpline<float>, TrajectoryTypes, float>;
+
+        /**
+         * Return initialized trajectories for trunk/foot ik cartesian with empty splines
+         */
+        inline Trajectories TrajectoriesInit() {
```
Comment (ysims) [inline module/motion/QuinticWalk/src/engine/TrajectoryUtils.hpp]: @Bidski @TrentHouliston if you have time can you have a quick look at this function and see if you can find a reason for memory issues?

File: module/motion/QuinticWalk/src/QuinticWalk.cpp [hunk 3]
```diff
@@ -0,0 +1,286 @@
...
+
+                // see if the walk engine has new goals for us
+                walkableState = walk_engine.updateState(dt, current_orders, walkableState);
+
+                if ((walk_engine.getState() != engine::WalkEngineState::IDLE) && walkableState) {  // todo
+                    calculateJointGoals();
+                }
+            }
+        });
+    }  // namespace motion
```
Comment (Bidski) [inline module/motion/QuinticWalk/src/QuinticWalk.cpp:197]:
```suggestion
    }
```

File: module/motion/QuinticWalk/src/QuinticWalk.cpp [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+
+    float QuinticWalk::getTimeDelta() {
+        // compute time delta depended if we are currently in simulation or reality
+        float dt;
+        auto current_time = NUClear::clock::now();
+        // only take real time difference if walking was not stopped before
+        // TODO Is this really necessary?
+        auto time_diff_ms = std::chrono::duration_cast<std::chrono::milliseconds>(current_time - last_update_time);
+        dt                = time_diff_ms.count() / 1000.0f;
```
Comment (Bidski) [inline module/motion/QuinticWalk/src/QuinticWalk.cpp:206]:
```suggestion
        auto current_time = NUClear::clock::now();
        // only take real time difference if walking was not stopped before
        // TODO Is this really necessary?
        float dt = std::chrono::duration_cast<std::chrono::milliseconds>(current_time - last_update_time).count() / 1000.0f;
```

File: module/motion/QuinticWalk/src/QuinticWalk.cpp [hunk 5]
```diff
@@ -0,0 +1,286 @@
...
+                                      )
+                .normalized()
+                .toRotationMatrix();
+        };
+
+        // Read the cartesian positions and orientations for trunk and fly foot
+        walk_engine.computeCartesianPosition(trunk_pos, trunk_axis, foot_pos, foot_axis, is_left_support);
+
+        // Change goals from support foot based coordinate system to trunk based coordinate system
+        Eigen::Affine3f Hst;  // trunk_to_support_foot_goal
```
Comment (Bidski) [inline module/motion/QuinticWalk/src/QuinticWalk.cpp:247]:
```suggestion
        Eigen::Affine3f Hst = Eigen::Affine3f::Identity();  // trunk_to_support_foot_goal
```

File: module/motion/QuinticWalk/src/QuinticWalk.cpp [hunk 6]
```diff
@@ -0,0 +1,286 @@
...
+        Hst.linear()      = setRPY(trunk_axis[0], trunk_axis[1], trunk_axis[2]).transpose();
+        Hst.translation() = -Hst.rotation() * trunk_pos;
+
+        Eigen::Affine3f Hfs;  // support_to_flying_foot
```
Comment (Bidski) [inline module/motion/QuinticWalk/src/QuinticWalk.cpp:251]:
```suggestion
        Eigen::Affine3f Hfs = Eigen::Affine3f::Identity();  // support_to_flying_foot
```

File: module/motion/QuinticWalk/src/engine/Footstep.cpp [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+#include <stdexcept>
+
+#include "utility/math/angle.h"
+
+namespace module {
+namespace motion {
+    namespace engine {
+
+        Footstep::Footstep(float foot_distance, bool is_left_support_foot) {
+            if (foot_distance <= 0.0) {
```
Comment (Bidski) [inline module/motion/QuinticWalk/src/engine/Footstep.cpp:18]:
```suggestion
            if (foot_distance <= 0.0f) {
```

File: module/motion/QuinticWalk/src/engine/TrajectoryUtils.hpp [hunk 8]
```diff
@@ -0,0 +1,308 @@
...
+         * assign it to given vector
+         */
+        inline void TrajectoriesTrunkFootPos(float t,
+                                             const Trajectories& traj,
+                                             Eigen::Vector3f& trunkPos,
+                                             Eigen::Vector3f& trunkAxis,
+                                             Eigen::Vector3f& footPos,
+                                             Eigen::Vector3f& footAxis) {
+            if (traj.size() != 14) {
+                NUClear::log(traj.size());
```
Comment (Bidski) [inline module/motion/QuinticWalk/src/engine/TrajectoryUtils.hpp:201]: Should probably make this say something more descriptive, and maybe make it a warning?

File: module/support/NUsight/data/config/NUsight.yaml [hunk 9]
```diff
@@ -12,7 +12,7 @@ reaction_handles:
   # Overview information for the dashboard, designed for minimum bandwidth
   overview: true
   # Motor positions, orientation, etc.
-  sensor_data: false
+  sensor_data: true
```
Comment (Bidski) [inline module/support/NUsight/data/config/NUsight.yaml:15]: Should probably revert the changes in this file

File: module/motion/QuinticWalk/src/QuinticWalk.cpp (hunk 4 above)
Comment (KipHamiltons) [reply module/motion/QuinticWalk/src/QuinticWalk.cpp:206]: Would making this change nullify the TODO above it?

File: module/motion/QuinticWalk/src/QuinticWalk.cpp (hunk 5 above)
Comment (KipHamiltons) [reply module/motion/QuinticWalk/src/QuinticWalk.cpp:247]: Is there a point to assigning it the identity if the rotation and linear are immediately set afterwards?

Comment (TrentHouliston) [reply module/motion/QuinticWalk/src/QuinticWalk.cpp:247]: I believe that setting the translation sets the right-hand column, and linear is the upper 3x3, but this might have something to do with the 0,0,0,1 at the bottom of the matrix? However, I think that the default constructor of Affine3f is already identity but you would have to confirm.

Comment (KipHamiltons) [reply module/motion/QuinticWalk/src/QuinticWalk.cpp:247]: based on the Eigen source, it looks like the default constructor sets the last row to `{0, 0, 0, 1}` automatically.
(https://gitlab.com/libeigen/eigen/-/blob/master/Eigen/src/Geometry/Transform.h line 258)

File: module/motion/QuinticWalk/src/QuinticWalk.cpp (hunk 4 above)
Comment (Bidski) [reply module/motion/QuinticWalk/src/QuinticWalk.cpp:206]: No, I think the TODO is about the calculation itself

File: module/motion/QuinticWalk/src/QuinticWalk.cpp (hunk 5 above)
Comment (Bidski) [reply module/motion/QuinticWalk/src/QuinticWalk.cpp:247]: I remember having issues in the kinematics code that required me to set it to the identity, but that may have been for a different reason

Comment (KipHamiltons) [reply module/motion/QuinticWalk/src/QuinticWalk.cpp:247]: I don't think it sets anything except the bottom row - I wonder if the other rows were causing the issues? do you remember which PR or files you had that problem?

Comment (TrentHouliston) [reply module/motion/QuinticWalk/src/QuinticWalk.cpp:247]: If it's only the 0,0,0,1 that's set then if you don't set the translation and linear it would have junk in it besides that final row. That could have been the problem that was occuring

Comment (Bidski) [reply module/motion/QuinticWalk/src/QuinticWalk.cpp:247]: Yea, I was using a running transform in ForwardKinematics, that was the issue. We should be fine without making this change.

File: module/motion/QuinticWalk/src/QuinticWalk.cpp [hunk 10]
```diff
@@ -140,30 +140,31 @@ namespace motion {
...
-                else if (std::abs(RPY[1]) > config.imu_pitch_threshold) {
-                    log<NUClear::WARN>(fmt::format("Robot pitch exceeds threshold - {} > {}",
-                                                   std::abs(RPY.y()),
-                                                   config.imu_pitch_threshold));
-                    walk_engine.requestPause();
-                }
+        imu_reaction = on<Trigger<Sensors>>().then([this](const Sensors& sensors) {
+            if (!config.imu_active) {
+                imu_reaction.disable();
+            }
```
Comment (Bidski) [inline module/motion/QuinticWalk/src/QuinticWalk.cpp:146]: Move this to `on<Configuration>`

File: module/support/NUsight/data/config/NUsight.yaml [hunk 11]
```diff
@@ -12,7 +12,7 @@ reaction_handles:
   # Overview information for the dashboard, designed for minimum bandwidth
   overview: true
   # Motor positions, orientation, etc.
-  sensor_data: true
+  sensor_data: false
```
Comment (Bidski) [inline module/support/NUsight/data/config/NUsight.yaml]: Probably doesn't need to be committed?

File: shared/tests/utility/motion/TestSplines.cpp [hunk 12]
```diff
@@ -0,0 +1,230 @@
...
+                                        {1, 0, 1},
+                                        {39, 21, 62359143990}};
+
+TEST_CASE("Test Combination", "[utility][motion][splines][Combination]") {
+    utility::motion::splines::Combination comb;
+
+    for (size_t i = 0; i < C.size(); i++) {
+        int n          = C.at(i).at(0);
+        int k          = C.at(i).at(1);
+        int nCk        = C.at(i).at(2);
```
Comment (Bidski) [inline shared/tests/utility/motion/TestSplines.cpp:136]:
```suggestion
        int n          = C[i][0];
        int k          = C[i][1];
        int nCk        = C[i][2];
```

File: shared/tests/utility/motion/TestSplines.cpp [hunk 13]
```diff
@@ -0,0 +1,230 @@
...
+        int two       = P.at(i).at(4);
+        int three     = P.at(i).at(3);
+        int four      = P.at(i).at(2);
+        int five      = P.at(i).at(1);
+        int six       = P.at(i).at(0);
+        int x         = P.at(i).at(6);
+        long int pos  = P.at(i).at(7);
+        long int vel  = P.at(i).at(8);
+        long int acc  = P.at(i).at(9);
+        long int jerk = P.at(i).at(10);
```
Comment (Bidski) [inline shared/tests/utility/motion/TestSplines.cpp:207]:
```suggestion
        int one       = P[i][5];
        int two       = P[i][4];
        int three     = P[i][3];
        int four      = P[i][2];
        int five      = P[i][1];
        int six       = P[i][0];
        int x         = P[i][6];
        long int pos  = P[i][7];
        long int vel  = P[i][8];
        long int acc  = P[i][9];
        long int jerk = P[i][10];
```

**Review by Bidski — CHANGES_REQUESTED** [review]

File: module/motion/QuinticWalk/src/QuinticWalk.cpp (hunk 10 above)
Comment (ysims) [reply module/motion/QuinticWalk/src/QuinticWalk.cpp:146]: It didn't work in `on<Configuration>` 🤷

Comment (TrentHouliston) [reply module/motion/QuinticWalk/src/QuinticWalk.cpp:146]: That's because you defined the reaction after the configuration block. So when you tried to disable it you hadn't even created the reaction yet.

Comment (ysims) [reply module/motion/QuinticWalk/src/QuinticWalk.cpp:146]: Ah that would be an issue yes

File: module/motion/QuinticWalk/src/QuinticWalk.cpp [hunk 14]
```diff
@@ -88,6 +113,10 @@ namespace motion {
                     jointGains[i] = cfg["gains"]["legs"].as<float>();
                 }
             }
+
+            if (!config.imu_active) {
+                imu_reaction.disable();
+            }
```
Comment (Bidski) [inline module/motion/QuinticWalk/src/QuinticWalk.cpp:119]:
```suggestion
            imu_reaction.enable(config.imu_active);
```

File: shared/tests/utility/motion/TestSplines.cpp [hunk 15]
```diff
@@ -127,19 +123,23 @@ std::vector<std::vector<long int>> C = {{3, 0, 1},
...
     utility::motion::splines::Combination comb;
 
+    // Loop over all our test values and see if Combination gives the correct result for n choose k
     for (size_t i = 0; i < C.size(); i++) {
-        int n          = C.at(i).at(0);
-        int k          = C.at(i).at(1);
-        int nCk        = C.at(i).at(2);
+        int n          = C[i][0];
+        int k          = C[i][1];
+        int nCk        = C[i][2];
```
Comment (Bidski) [inline shared/tests/utility/motion/TestSplines.cpp:135]:
```suggestion
        const int& n          = C[i][0];
        const int& k          = C[i][1];
        const int& nCk        = C[i][2];
```

File: shared/tests/utility/motion/TestSplines.cpp [hunk 16]
```diff
@@ -191,40 +191,120 @@ std::vector<std::vector<long int>> P = {{3, 3, 2, 2, 0, 5, 11, 529985, 236357, 8
...
+            std::vector<double> splineResult = spline.part(i).polynom.getCoefs();
+
+            // Check contraints for first endpoint and then the second endpoint in the polynomial
+            for (int k = 0; k < 2; k++) {
+
+                // We need to check what this spline actually gives for x, and the first and second derivatives
+                // of x, given an input of t. We will then check they match the expected values.
+                double x   = 0;
+                double xd  = 0;
+                double xdd = 0;
```
Comment (Bidski) [inline shared/tests/utility/motion/TestSplines.cpp:278]:
```suggestion
                double x   = 0.0;
                double xd  = 0.0;
                double xdd = 0.0;
```

File: shared/tests/utility/motion/TestSplines.cpp [hunk 17]
```diff
@@ -191,40 +191,120 @@ std::vector<std::vector<long int>> P = {{3, 3, 2, 2, 0, 5, 11, 529985, 236357, 8
...
 
-        int long poly_pos  = poly.pos(x);
-        int long poly_vel  = poly.vel(x);
-        int long poly_acc  = poly.acc(x);
-        int long poly_jerk = poly.jerk(x);
+                // Each spline is not made using the given t values. Each spline starts at t = 0 and the next point is
+                // defined as t = second-first
+                double t = k == 0 ? 0 : points[i + 1][0] - points[i][0];
```
Comment (Bidski) [inline shared/tests/utility/motion/TestSplines.cpp:282]:
```suggestion
                double t = ((k == 0) ? 0.0 : points[i + 1][0] - points[i][0]);
```

File: shared/tests/utility/motion/TestSplines.cpp [hunk 18]
```diff
@@ -191,40 +191,120 @@ std::vector<std::vector<long int>> P = {{3, 3, 2, 2, 0, 5, 11, 529985, 236357, 8
...
 
-        INFO("p(x) = " << one << "x^5 + " << two << "x^4 + " << three << "x^3 + " << four << "x^2 + " << five << "x + "
-                       << six << ".");
+                // Loop over the coefficients, add to the calculations for each coefficient.
+                // Position: x += a_j t^j
+                // Velocity: xd = j * a_j * t^(j-1)
+                // Acceleration: xdd = j * (j-1) * a_j * t^(j-2)
+                // If j is less than 0, don't add anything since our derivative has become 0 for that term.
+                for (int j = 0; j < splineResult.size(); j++) {
```
Comment (Bidski) [inline shared/tests/utility/motion/TestSplines.cpp:289]:
```suggestion
                for (int j = 0; j < int(splineResult.size()); j++) {
```

File: shared/tests/utility/motion/TestSplines.cpp [hunk 19]
```diff
@@ -191,40 +191,120 @@ std::vector<std::vector<long int>> P = {{3, 3, 2, 2, 0, 5, 11, 529985, 236357, 8
...
+                    x += splineResult[j] * pow(t, j);
+                    xd += (j - 1) >= 0 ? j * splineResult[j] * pow(t, j - 1) : 0;
+                    xdd += (j - 2) >= 0 ? j * (j - 1) * splineResult[j] * pow(t, j - 2) : 0;
```
Comment (Bidski) [inline shared/tests/utility/motion/TestSplines.cpp:292]:
```suggestion
                    xd += (((j - 1) >= 0) ? j * splineResult[j] * pow(t, j - 1) : 0.0);
                    xdd += (((j - 2) >= 0) ? j * (j - 1) * splineResult[j] * pow(t, j - 2) : 0.0);
```

**Review by Bidski — CHANGES_REQUESTED** [review]

**Review by Bidski — APPROVED** [review]
LGTM ..... assuming it works

File: module/behaviour/planning/SimpleWalkPathPlanner/src/SimpleWalkPathPlanner.cpp [hunk 20]
```diff
@@ -201,9 +201,9 @@ namespace behaviour {
 
                     arma::vec3 rBWw_temp = {ball.position[0], ball.position[1], fieldDescription.ball_radius};
                     rBWw                 = timeSinceBallSeen < search_timeout ? rBWw_temp :  // Place last seen
-                               Htw.i().x() + Htw.i().translation();          // In front of the robot
-                    arma::vec3 pos       = Htw.transformPoint(rBWw);
-                    position             = pos.rows(0, 1);
+                               Htw.i().x() + Htw.i().translation();                          // In front of the robot
+                    arma::vec3 pos = Htw.transformPoint(rBWw);
```
Comment (Bidski) [inline module/behaviour/planning/SimpleWalkPathPlanner/src/SimpleWalkPathPlanner.cpp:205]: :nauseated_face:

**Review by Bidski — CHANGES_REQUESTED** [review]
There seem to be a number of formatting changes to files that have nothing to do with this PR

File: module/behaviour/planning/SimpleWalkPathPlanner/src/SimpleWalkPathPlanner.cpp (hunk 20 above)
Comment (ysims) [reply module/behaviour/planning/SimpleWalkPathPlanner/src/SimpleWalkPathPlanner.cpp:205]: Yeah this is what happened when I ran `./b format`. Then it failed again. Then I realised I haven't got the new docker image on here. Going through LLVM rn :laughing: Will fix it up when that's finished.

Comment (KipHamiltons) [reply module/behaviour/planning/SimpleWalkPathPlanner/src/SimpleWalkPathPlanner.cpp:205]: I was about to message you asking if you'd updated your container :laughing:

**Review by Bidski — CHANGES_REQUESTED** [review]
There are still changes to unrelated files. Perhaps we need to make another PR that updates all of the formatting

**Comment by ysims** [comment]
Which files? I can't see any changes to unrelated files?

**Comment by Bidski** [comment]
Oh, nevermind, your last commit edited unrelated files, but you were just undoing the changes you made before

**Review by Bidski — APPROVED** [review]
OK, looks good now
